image-mdraid.c

```diff
@@ -0,0 +1,414 @@
+		isuper++;
+	}
+
+	if (size == 2)
+		newcsum += __le16_to_cpu(*(unsigned short*) isuper);
+
+	csum = (newcsum & 0xffffffff) + (newcsum >> 32);
+	sb->sb_csum = disk_csum; //Set checksum in struct back to original value
+	return __cpu_to_le32(csum);
+}
+
+static int mdraid_generate(struct image *image) {
+	mdraid_img_t *md = image->handler_priv;
+	mdraid_img_t *mdp = NULL; //Inheriting from this parent if not NULL
+	__le16 max_devices;
+
+	//Determine max_devices while considering possibility of inheritance from other image
+	if (md->img_parent) {
+		mdp = md->img_parent->handler_priv;
+		max_devices = mdp->sb->raid_disks;
+	} else {
+		max_devices = cfg_getint(image->imagesec, "devices");
+	}
+
+	//Determine role of this device in array
+	__le16 role = cfg_getint(image->imagesec, "role");
+	if (cfg_getint(image->imagesec, "role") == -1) { //If role is -1 it should be autoassigned to parenting devices
+		if (mdp) {
+			role = ++mdp->last_role; //Take role from master and increment its counter
+		} else {
+			role = 0; //Master has role of 0
+		}
+		image_info(image, "MDRAID automaticaly assigned role %d.\n", role);
+	}
+
+	if (role > MD_DISK_ROLE_MAX) {
+		image_error(image, "MDRAID role has to be >= 0 and <= %d.\n", MD_DISK_ROLE_MAX);
+		return 6;
+	}
+
+	if (role >= max_devices) {
+		image_error(image, "MDRAID role of this image has to be lower than total number of %d devices (roles are counted from 0).\n", max_devices);
+		return 5;
+	}
+
+	//MD Superblock and Bitmap Superblock
+	size_t superblock_size = sizeof(struct mdp_superblock_1) + max_devices*2;
+	struct mdp_superblock_1 *sb = md->sb = xzalloc(superblock_size);
+	bitmap_super_t *bsb = &md->bsb;
+
+	if (mdp) {
+		//We are inheriting the superblock in this case
+		memcpy(md->sb,  mdp->sb,  superblock_size);
+		//memcpy(&md->bsb, &mdp->bsb, sizeof(bitmap_super_t));
+	} else {
+		//We are not inheriting superblock, therefore we need to fully initialize the array
+
+		char *name = cfg_getstr(image->imagesec, "label");
+
+		/* constant array information - 128 bytes */
+		sb->magic = MD_SB_MAGIC;	/* MD_SB_MAGIC: 0xa92b4efc - little endian. */
+		sb->major_version = 1;	/* 1 */
+		sb->feature_map = MD_FEATURE_BITMAP_OFFSET;	/* bit 0 set if 'bitmap_offset' is meaningful */
+		sb->pad0 = 0;		/* always set to 0 when writing */
+
+		char *raid_uuid = cfg_getstr(image->imagesec, "raid-uuid");
+		if (!raid_uuid) raid_uuid = uuid_random();
+		uuid_parse(raid_uuid, sb->set_uuid);  /* user-space generated. U8[16]*/
+
+		strncpy(sb->set_name, name, 32); sb->set_name[31] = 0;	/* set and interpreted by user-space. CHAR[32] */
+
+		long int timestamp = cfg_getint(image->imagesec, "timestamp");
+		if (timestamp >= 0) {
+			sb->ctime = timestamp & 0xffffffffff;
+		} else {
+			sb->ctime = mdraid_time & 0xffffffffff;	/* lo 40 bits are seconds, top 24 are microseconds or 0*/
+		}
+
+		sb->level = 1;		/* -4 (multipath), -1 (linear), 0,1,4,5 */
+		//sb->layout = 2;		/* only for raid5 and raid10 currently */
```

Comment:
This is confusing. Is the value a don't care for other raid levels? In that case, maybe use
```
// sb->layout = ignored;
```
Or something else that would not compile.

Reply:
Can do that, BUT. Are we planning to (un)comment random lines? I mean... the code surely is not gonna work if you comment the line before that, but still will compile. Where do we draw the line here?

Reply:
BTW the array is going to be completely valid and fine if you uncomment that line, since every implementation knows to ignore this field for RAID 1. What risk exactly are you trying to avoid?

Reply:
I'm trying to avoid confusion. Where is the '2' coming from?